File: app/src/main/java/org/arcade/atomcity/presentation/viewmodel/TaikoViewModel.kt
package org.arcade.atomcity.presentation.viewmodel

import android.annotation.SuppressLint
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import org.arcade.atomcity.model.taikoserver.usersettings.TaikoServerUserSettingsResponse
import kotlinx.coroutines.launch
import org.arcade.atomcity.domain.usecase.GetTaikoServerDataUseCase
import org.arcade.atomcity.model.taikoserver.musicDetails.TaikoServerMusicDetailsResponse
import org.arcade.atomcity.model.taikoserver.songHistory.TaikoServerPlayHistoryResponse
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import java.net.URL
import androidx.core.graphics.createBitmap
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class TaikoViewModel(private val usecase: GetTaikoServerDataUseCase) : ViewModel() {

    // StateFlow to hold the music details data
    private val _scoresData = MutableStateFlow<TaikoServerPlayHistoryResponse?>(null)
    val scoresData = _scoresData

    // StateFlow to hold the music details data
    private val _musicDetailsData = MutableStateFlow<TaikoServerMusicDetailsResponse?>(null)
    val musicDetailsData = _musicDetailsData

    // StateFlow to hold the user settings data
    private val _userSettingsData = MutableStateFlow<TaikoServerUserSettingsResponse?>(null)
    val userSettingsData = _userSettingsData

    val isLoading = MutableStateFlow(false)
    val isLoadingMusicDetails = MutableStateFlow(false)
    val isLoadingUserSettings = MutableStateFlow(false)
    val isLoadingScores = MutableStateFlow(false)

    internal val _currentPage = MutableStateFlow(1)

    fun onPageChange(newPage: Int) {
        _currentPage.value = newPage
    }

    suspend fun fetchPlayHistoryPlayData(page: Int) {
        isLoadingScores.value = true
        try {
            usecase.getPlayHistoryFlow(page.toString()).collect { response ->
                _scoresData.value = response
            }
        } catch (e: Exception) {
            Log.e("TaikoScoresViewModel", "Error: ${e.message}")
            isLoadingScores.value = false
        } finally {
            isLoadingScores.value = false
        }
    }

    suspend fun fetchMusicDetails() {
        isLoadingMusicDetails.value = true
        try {
            usecase.getMusicDetailsFlow().collect { response ->
                _musicDetailsData.value = response
            }
        } catch (e: Exception) {
            Log.e("TaikoScoresViewModel", "Error: ${e.message}")
            isLoadingMusicDetails.value = false
        } finally {
            isLoadingMusicDetails.value = false
        }
    }

    suspend fun getUserSettings() {
        isLoadingUserSettings.value = true
        try {
            usecase.getUserSettingsFlow().collect { response ->
                _userSettingsData.value = response
            }
        } catch (e: Exception) {
            Log.e("TaikoViewModel", "Error fetching user settings: ${e.message}")
            isLoadingUserSettings.value = false
        } finally {
            isLoadingUserSettings.value = false
        }
    }

    fun mergeMusicDetailsWithScores() {
        val scores = scoresData.value
        val musicDetails = musicDetailsData.value

        val mergedData = scores!!.copy(
            taikoServerSongHistoryData = scores.taikoServerSongHistoryData.map { score ->
                val musicDetail = musicDetails?.entries?.find { it.key == score.songId.toString() }?.value
                if (musicDetail != null) {
                    score.copy(
                        musicName = musicDetail.songName,
                        musicArtist = musicDetail.artistName
                    )
                } else {
                    score
                }
            }.reversed()
        )
        _scoresData.value = mergedData
    }


    fun getScores() {
        viewModelScope.launch {
            fetchMusicDetails()
            fetchPlayHistoryPlayData(1)
            getUserSettings()
            mergeMusicDetailsWithScores()
            getAvatar()
        }
    }

    @SuppressLint("DefaultLocale")
    fun getAvatar() {
        viewModelScope.launch {
            val kigurumi = userSettingsData.value?.kigurumi?.let { String.format("%04d", it) } ?: "0000"
            val head = userSettingsData.value?.head?.let { String.format("%04d", it) } ?: "0000"
            val body = userSettingsData.value?.body?.let { String.format("%04d", it) } ?: "0000"
            val face = userSettingsData.value?.face?.let { String.format("%04d", it) } ?: "0000"
            val puchi = userSettingsData.value?.puchi?.let { String.format("%04d", it) } ?: "0000"

            val bodyUrl = "https://taiko.farewell.dev/images/Costumes/body/body-$body.webp"
            val headUrl = "https://taiko.farewell.dev/images/Costumes/head/head-$head.webp"
            val faceUrl = "https://taiko.farewell.dev/images/Costumes/face/face-$face.webp"
            val puchiUrl = "https://taiko.farewell.dev/images/Costumes/puchi/puchi-$puchi.webp"

            val avatarUrls = listOf(bodyUrl, headUrl, faceUrl, puchiUrl)
            loadMergedAvatar(avatarUrls)
        }
    }

    val mergedAvatar = MutableStateFlow<Bitmap?>(null)

    fun loadMergedAvatar(urls: List<String>) {
        viewModelScope.launch {
            val bitmap = mergeImages(urls)
            mergedAvatar.value = bitmap
        }
    }

    suspend fun mergeImages(urls: List<String>): Bitmap = withContext(Dispatchers.IO) {
        val bitmaps: List<Bitmap> = urls.map { url ->
            val input = URL(url).openStream()
            BitmapFactory.decodeStream(input)
        }
        if (bitmaps.isEmpty()) {
            throw IllegalArgumentException("No bitmaps provided to merge.")
        }
        val result: Bitmap = createBitmap(
            bitmaps[0].width,
            bitmaps[0].height,
            bitmaps[0].config ?: Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(result)
        bitmaps.forEach { bitmap -> canvas.drawBitmap(bitmap, 0f, 0f, null) }
        result
    }
}

File: iosApp/iosApp/TaikoViewModel.swift
//
//  TaikoViewModel.swift
//  iosApp
//
//  Created by Mohamed Zidani on 10/08/2026.
//

import Foundation
import shared
import SwiftUI

class TaikoViewModel: ObservableObject {
    private let useCase: GetTaikoServerDataUseCase = KoinProxy.shared.getTaikoUseCase()
    private let apiKeyManager: ApiKeyManager = KoinProxy.shared.getApiKeyManager()
    
    @Published var scores: [TaikoServerSongHistoryData] = []
    @Published var userSettings: TaikoServerUserSettingsResponse? = nil
    @Published var avatar: UIImage? = nil
    @Published var isLoading: Bool = false
    
    func getScores() {
        guard let userIdString = apiKeyManager.getApiKey(gameName: "taiko"),
              let userId = Int32(userIdString) else {
            print("No valid user ID found for Taiko")
            return
        }
        
        self.isLoading = true
        
        Task {
            // Sequential fetch to match Android logic
            await fetchUserSettings(userId: userId)
            await fetchPlayHistory(userId: userId)
            await getAvatar()
            
            DispatchQueue.main.async {
                self.isLoading = false
            }
        }
    }
    
    private func fetchUserSettings(userId: Int32) async {
        return await withCheckedContinuation { continuation in
            useCase.getUserSettingsFlow(userNumber: String(userId)).collect(collector: Collector<TaikoServerUserSettingsResponse?> { response in
                DispatchQueue.main.async {
                    self.userSettings = response
                    continuation.resume()
                }
            }) { _ in }
        }
    }
    
    private func fetchPlayHistory(userId: Int32) async {
        // We also need music details to merge names
        let musicDetails: TaikoServerMusicDetailsResponse? = await withCheckedContinuation { continuation in
            useCase.getMusicDetailsFlow().collect(collector: Collector<TaikoServerMusicDetailsResponse?> { response in
                continuation.resume(returning: response)
            }) { _ in }
        }
        
        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
            useCase.getPlayHistoryFlow(userNumber: String(userId)).collect(collector: Collector<TaikoServerPlayHistoryResponse?> { response in
                if let response = response {
                    var merged = response.taikoServerSongHistoryData
                    if let details = musicDetails?.entries {
                        merged = merged.map { score in
                            if let detail = details[String(score.songId)] {
                                score.doCopy(
                                    id: score.id,
                                    songId: score.songId,
                                    difficulty: score.difficulty,
                                    score: score.score,
                                    bad: score.bad,
                                    good: score.good,
                                    great: score.great,
                                    maxCombo: score.maxCombo,
                                    playTime: score.playTime,
                                    drumroll: score.drumroll,
                                    musicName: detail.songName,
                                    musicArtist: detail.artistName
                                )
                            } else {
                                score
                            }
                        }.reversed()
                    }
                    DispatchQueue.main.async {
                        self.scores = merged
                        continuation.resume()
                    }
                } else {
                    continuation.resume()
                }
            }) { _ in }
        }
    }
    
    private func getAvatar() async {
        guard let settings = userSettings else { return }
        
        let kigurumi = String(format: "%04d", settings.kigurumi?.int32Value ?? 0)
        let head = String(format: "%04d", settings.head?.int32Value ?? 0)
        let body = String(format: "%04d", settings.body?.int32Value ?? 0)
        let face = String(format: "%04d", settings.face?.int32Value ?? 0)
        let puchi = String(format: "%04d", settings.puchi?.int32Value ?? 0)
        
        let urls = [
            "https://taiko.farewell.dev/images/Costumes/body/body-\(body).webp",
            "https://taiko.farewell.dev/images/Costumes/head/head-\(head).webp",
            "https://taiko.farewell.dev/images/Costumes/face/face-\(face).webp",
            "https://taiko.farewell.dev/images/Costumes/puchi/puchi-\(puchi).webp"
        ]
        
        let images = await withTaskGroup(of: UIImage?.self) { group in
            for urlString in urls {
                group.addTask {
                    guard let url = URL(string: urlString),
                          let data = try? Data(contentsOf: url) else { return nil }
                    return UIImage(data: data)
                }
            }
            
            var collected: [UIImage] = []
            for await image in group {
                if let image = image {
                    collected.append(image)
                }
            }
            return collected
        }
        
        if !images.isEmpty {
            let size = images[0].size
            let renderer = UIGraphicsImageRenderer(size: size)
            let merged = renderer.image { _ in
                for image in images {
                    image.draw(in: CGRect(origin: .zero, size: size))
                }
            }
            DispatchQueue.main.async {
                self.avatar = merged
            }
        }
    }
}
